staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go

```diff
@@ -99,6 +182,56 @@ type waitFor struct {
 	readyAt time.Time
 	// index in the priority queue (heap)
 	index int
+	// options defines how the caller would like the item to behave.
+	options DelayingOptions
+	// features describes additional waitingLoop features we want to apply.
+	features waitingLoopFeatures
+	// returnCh allows the caller to block pending the processing and to also return
+	// an int value which can represent a stat such as existence or 'waiting' queue length.
+	returnCh chan int
+}
+
+// Funtions to render DelayingOptions in a more human concise human-consumable form.
+// String satisfies the Stringer interface to allow this representation to be used anywhere %s is used.
+func (w waitFor) String() string {
+	if w.features.forget || w.features.check_existence {
+		return fmt.Sprintf("%s %v", w.features, w.data)
+	} else if w.features.check_length {
+		return w.features.String()
+	}
+	return fmt.Sprintf("%s:%v Ready:%s Options:{%s}", w.features, w.data, format_time(w.readyAt), w.options)
+}
+
+// waitingLoopFeatures are the implementation features provided by the waitingLoop routine which are not directly
+// accessible to clients of the queue.  The features can only be provided from within the waitingLoop because it has
+// exclusive access to the state of the 'waiting' queue.
+type waitingLoopFeatures struct {
+	// forget - request the waitingLoop forget an item in the 'waiting' queue.
+	forget bool
+	// synchronous - used internally
+	// force calls to wait for processing before returning e.g. Add, AddAfter, AddWithOptions, Forget will wait
+	// for the item to make its way through the waitingLoopAddCh queue, get picked up by the waitingLoop and processed into the
+	// appropriate queue.  This makes it safe to immediately assert condtions about the queue for testing, but it is also used
+	// to emulate the original Add() and AddAfter() behaviour (when the duration is 0).
+	synchronous bool
+	// check_existence request the waitingLoop return an integer on the returnCh, 0/1 to represent the presence of the item
+	// in the 'waiting' queue.
+	check_existence bool
+	// check_length request the waitingLoop to return the length of the 'waiting' queue as an integer on the returnCh.
+	check_length bool
```

Comment:
Yes, the problem is that these must be supported by the messages and processed from within the waitingLoop - otherwise I would have tried to insert them into the test code instead.  I couldn't find a way to do that.  If I swapped out the waitingLoop somehow then the tests wouldn't be testing the actual implementation code.